client/packages/grandpa-ranger/src/index.ts

```diff
@@ -0,0 +1,138 @@
+import {generateRange} from "./collect";
+
+require('dotenv').config()
+import { Connection } from './connection';
+import { cryptoWaitReady } from "@t3rn/sdk"
+import { Prometheus } from "./prometheus";
+
+class GrandpaRanger {
+	circuit: Connection;
+	target: Connection;
+	config: any;
+	prometheus: Prometheus;
+
+	constructor(config: any) {
+		this.config = config;
+		this.prometheus = new Prometheus();
+		this.prometheus.rangeBreak.inc(this.config.rangeBreak);
+		this.prometheus.nextSubmission.set(Date.now() + this.config.rangeBreak * 1000);
+	}
+
+	async start() {
+		await this.connectClients();
+		this.scheduleRangeSubmission();
+	}
+
+	async connectClients() {
+		await cryptoWaitReady()
+		this.circuit = new Connection(this.config.circuit.rpc1, this.config.circuit.rpc2, true, this.prometheus, this.config.circuitSigner);
+		this.circuit.connect();
+		this.target = new Connection(this.config.target.rpc1, this.config.target.rpc2, false, this.prometheus);
+		this.target.connect();
+	}
+
+	async collectAndSubmit(resolve: any) {
+		if (!this.circuit.isActive || !this.target.isActive) return resolve() // skip if either client is not active
+
+		let batches = await generateRange(this.config, this.circuit, this.target, this.prometheus)
+			.catch((e) => {
+				console.log(e);
+				// potentially we want to introduce a retry logic here
+				return resolve()
+			})
+
+		if(batches.length > 0) {
+			// calculate the total number of elements in the batches elements
+			const totalElements = batches.reduce((acc, curr) => acc + curr.range.length, 0)
+
+			this.submitToCircuit(batches)
+				.then((res) => {
+					console.log({"status": "Submitted", "range_size": totalElements, "circuit_block": res})
+					this.prometheus.nextSubmission.set(Date.now() + this.config.rangeBreak * 1000);
+					this.prometheus.successes.inc({rangeSize: totalElements, circuitBlock: res, timestamp: Date.now()})
```

Comment:
timestamp, circuitblock and rangesize shouldnt be used as labels
https://prometheus.io/docs/practices/instrumentation/#do-not-overuse-labels

circuitblock should be different metric if its not the same as circuitHeight